kotlin-ktor/src/test/java/com/baeldung/graphql/client/GraphQLClientTest.kt

```diff
@@ -0,0 +1,142 @@
+package com.baeldung.graphql.client
+
+import com.baeldung.graphql.server.data.AttendeeRepository
+import com.baeldung.graphql.server.data.ConferenceRepository
+import com.baeldung.graphql.server.graphQLModule
+import io.ktor.server.engine.*
+import io.ktor.server.netty.*
+import kotlinx.coroutines.runBlocking
+import org.junit.AfterClass
+import org.junit.BeforeClass
+import kotlin.test.*
+
+class GraphQLClientTest {
+
+    private val graphQLService = GraphQLService("http://0.0.0.0:8080/graphql")
+
+    // https://stackoverflow.com/questions/61245981/how-can-i-start-ktor-server-programmatically-for-integration-tests
+    // Since regular Ktor tests work with a dummy application with no server, we can't actually test the client properly without starting it
+    // So we start it up and tear it down for every test.
+    companion object {
+
+        private lateinit var server: NettyApplicationEngine
+
+        @BeforeClass
+        @JvmStatic
+        fun setup() {
+            val env = applicationEngineEnvironment {
+                module {
+                    graphQLModule()
+                }
+                connector {
+                    host = "0.0.0.0"
+                    port = 8080
+                }
+            }
+            server = embeddedServer(Netty, env).start(false)
+        }
+
+        @AfterClass
+        @JvmStatic
+        fun teardown() {
+            // clean up after this class, leave nothing dirty behind
+            server.stop(1000, 10000)
+        }
+    }
+
+    @BeforeTest
+    fun cleanUp() {
+        ConferenceRepository.clear()
+        AttendeeRepository.clear()
+    }
+
+    private suspend fun createDummyConference(attendees: List<Int> = listOf()) {
+        graphQLService.createOrSaveConference(null, "Some Conference", attendees)
+    }
+
+    private suspend fun createDummyAttendee() {
+        graphQLService.createOrSaveAttendee(null, "Some Attendee")
+    }
+
+    @Test
+    fun `given conference by id when fetching existing conference then should fetch it`(): Unit = runBlocking {
+        createDummyConference()
+        val fetchedConference = graphQLService.findConferenceById(0)
+        assertEquals(0, fetchedConference?.id)
+        assertEquals("Some Conference", fetchedConference?.name)
+    }
+
+    @Test
+    fun `given conference by id when fetching not existing conference then should return null`(): Unit = runBlocking {
+        val fetchedConference = graphQLService.findConferenceById(0)
+        assertNull(fetchedConference)
+    }
+
+    @Test
+    fun `given create or save conference when saving not existing conference then should create it`(): Unit = runBlocking {
+        graphQLService.createOrSaveConference(null, "Some Conference")
+        val createdConference = ConferenceRepository.findById(0)
+        assertEquals("Some Conference", createdConference?.name)
+    }
```

Comment:
What about the freshly assigned conference id? Is it correctly returned in the response to the client?

Reply:
✅ Good point, missed that test case